fpfilter.pl

```diff
@@ -603,3 +604,128 @@ sub index_bam {
     system(@args) == 0
         or die "Unable to index $bam: $?\n";
 }
+
+sub validate_file_for_reading {
+    my $file = shift;
+
+    unless ( defined $file ) {
+        Carp::croak("Can't validate_file_for_reading: No file given");
+    }
+
+    if ($file eq '-') {
+        return 1;
+    }
+
+    unless (-e $file ) {
+        Carp::croak("File ($file) does not exist");
+    }
+
+    unless (-f $file) {
+        Carp::croak("File ($file) exists but is not a plain file");
+    }
+
+    unless ( -r $file ) {
+        Carp::croak("Do not have READ access to file ($file)");
+    }
+
+    return 1;
+}
+
+# Follows a symlink chain to reach the final file, accounting for relative symlinks along the way
+sub follow_symlink {
+    my $file = shift;
+
+    # Follow the chain of symlinks
+    while (-l $file) {
+        my $original_file = $file;
+        $file = readlink($file);
+        # If the symlink was relative, repair that
+        unless (File::Spec->file_name_is_absolute($file)) {
+            my $path = dirname($original_file);
+            $file = join ("/", ($path, $file));
+        }
+        validate_file_for_reading($file);
+    }
+
+    return $file;
+}
+
+sub file_type {
+    my $file = shift;
+
+    validate_file_for_reading($file);
+    $file = follow_symlink($file);
+
+    my $result = `file -b $file`;
+    my @answer = split /\s+/, $result;
+    return $answer[0];
+}
+
+sub file_is_gzipped {
+    my $filename = shift;
+
+    my $file_type = file_type($filename);
+
+    #NOTE: debian bug #522441 - `file` can report gzip files as any of these....
+    if ($file_type eq "gzip" or $file_type eq "Sun" or $file_type eq "Minix" or $file_type eq 'GRand') {
+        return 1;
+    } else {
+        return 0;
+    }
+}
+
+sub _open_file {
+    my ($file, $rw) = @_;
+    if ($file eq '-') {
+        if ($rw eq 'r') {
+            return 'STDIN';
+        }
+        elsif ($rw eq 'w') {
+            return 'STDOUT';
+        }
+        else {
+            die "cannot open '-' with access '$rw': r = STDIN, w = STDOUT!!!";
+        }
+    }
+    my $fh = (defined $rw) ? IO::File->new($file, $rw) : IO::File->new($file);
+    return $fh if $fh;
+    Carp::croak("Can't open file ($file) with access '$rw': $!");
+}
+
+sub open_gzip_file_for_reading {
+    my $file = shift;
+
+    validate_file_for_reading($file)
+        or return;
+
+    unless (file_is_gzipped($file)) {
+        Carp::croak("File ($file) is not a gzip file");
+    }
+
+    my $pipe = "zcat ".$file." |";
+
+    # _open_file throws its own exception if it doesn't work
+    return _open_file($pipe);
+}
+
+sub open_file_for_reading {
+    my $file = shift;
+
+    validate_file_for_reading($file)
+        or return;
+
+    # _open_file throws its own exception if it doesn't work
+    return _open_file($file, 'r');
```

Comment:
It ought to. `_open_file` ends up on [this line](https://github.com/jasonwalker80/fpfilter-tool/blob/c901a8a168167ae4ad7db56229c3e3ad4de5e902/fpfilter.pl#L690):
```perl
my $fh = (defined $rw) ? IO::File->new($file, $rw) : IO::File->new($file);
```
so we could do `IO::File->new($file, 'r')` to match.